src/mame/taito/pkspirit.cpp

```diff
@@ -83,64 +81,111 @@ class pkspirit_state : public driver_device
 	required_device<te7750_device> m_mainio;
 	required_device<gfxdecode_device> m_gfxdecode;
 	required_device<palette_device> m_palette;
+	required_shared_ptr<u16> m_vidram;
 
-	required_shared_ptr<uint16_t> m_bg_videoram;
-	required_shared_ptr<uint16_t> m_fg_videoram;
 	required_memory_bank m_audiobank;
 
 	required_device_array<taitoio_opto_device, 2> m_opto;
 
-	tilemap_t *m_bg_tilemap = nullptr;
-	tilemap_t *m_fg_tilemap = nullptr;
+	void vidram_w(offs_t offset, u16 data, u16 mem_mask);
 
 	uint32_t screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
 
-	void bg_videoram_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
-	void fg_videoram_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
-
-	TILE_GET_INFO_MEMBER(get_bg_tile_info);
-	TILE_GET_INFO_MEMBER(get_fg_tile_info);
-
 	void main_map(address_map &map);
 	void sound_map(address_map &map);
 };
 
-void pkspirit_state::bg_videoram_w(offs_t offset, uint16_t data, uint16_t mem_mask)
-{
-	COMBINE_DATA(&m_bg_videoram[offset]);
-	m_bg_tilemap->mark_tile_dirty(offset);
-}
-
-TILE_GET_INFO_MEMBER(pkspirit_state::get_bg_tile_info)
-{
-	int tileno = m_bg_videoram[tile_index];
-	tileinfo.set(0, tileno, 0x8 + 0x6, 0);
-}
-
-void pkspirit_state::fg_videoram_w(offs_t offset, uint16_t data, uint16_t mem_mask)
-{
-	COMBINE_DATA(&m_fg_videoram[offset]);
-	m_fg_tilemap->mark_tile_dirty(offset);
-}
-
-TILE_GET_INFO_MEMBER(pkspirit_state::get_fg_tile_info)
-{
-	int tileno = m_fg_videoram[tile_index];
-	tileinfo.set(0, tileno, 0x8 + 0x6, 0);
-}
-
 
 void pkspirit_state::video_start()
 {
-	m_bg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(pkspirit_state::get_bg_tile_info)), TILEMAP_SCAN_ROWS, 16, 16, 64, 32);
-	m_fg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(pkspirit_state::get_fg_tile_info)), TILEMAP_SCAN_ROWS, 16, 16, 64, 32);
-	m_fg_tilemap->set_transparent_pen(0);
 }
 
 uint32_t pkspirit_state::screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
 {
-	m_bg_tilemap->draw(screen, bitmap, cliprect, 0, 0);
-	m_fg_tilemap->draw(screen, bitmap, cliprect, 0, 0);
+	bitmap.fill(0, cliprect);
+
+	u32 spriteram_start = 0x300ff2; // should be 0xb10000 but a DMA is missing
+	u32 spriteram2_start = 0xb00000;
+
+	for (int i = 0; i < 0x800; i += 8)
+	{
+		// somewhat illogical, but the first sprite (one of the large tilemap sized sprites) seems to need to be drawn last
+		// (or there is some priority control)
+		int reali = (i + 0x8) & 0x7ff;
+
+		uint16_t sp0 = m_maincpu->space().read_word(spriteram_start + reali + 0);
```

Comment:
Using `space()` like this won't work when the actual DMA support is added later. Consider using `required_shared_ptr` for all the relevant buffers instead.

Reply:
it will work fine? the only thing the DMA is doing is copying the list from one place to another, change the address to the destination as per the comment and the code will still function.

Reply:
I don't mean to make the actual DMA copy but rather have some explicit pointers. Using `m_maincpu->space().read_word` N times is also inefficient when it comes to performance.

Reply:
seems like splitting hairs, as it will need changing once the core does DMA, and performance diff here is negligible, but I've changed it to a pointer directly into mainram instead.